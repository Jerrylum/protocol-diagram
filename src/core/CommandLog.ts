import { makeAutoObservable } from "mobx";

export type LogLevel = "info" | "error";

export interface CommandLog {
  level: LogLevel;
  message: string;
}

export class CommandLogger {
  private _logs: CommandLog[] = [];
  private _logCount = 0;

  constructor(public _maxLogs: number = 100) {
    makeAutoObservable(this);
  }

  get logs(): CommandLog[] {
    return this._logs;
  }

  get logCount(): number {
    return this._logCount;
  }

  add(level: LogLevel, message: string) {
    this._logs.push({ level, message });
    this._logs = this._logs.slice(-this._maxLogs);
    this._logCount++;
  }

  info(message: string) {
    this.add("info", message);
  }

  error(message: string) {
    this.add("error", message);
  }

  clear() {
    this._logs = [];
  }

  get maxLogs(): number {
    return this._maxLogs;
  }

  set maxLogs(maxLogs: number) {
    this._maxLogs = maxLogs;
    this._logs = this._logs.slice(-maxLogs);
  }
}
