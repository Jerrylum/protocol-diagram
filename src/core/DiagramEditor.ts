import { makeAutoObservable } from "mobx";
import { clamp } from "./Util";
import { Vector } from "./Vector";

export class DiagramEditor {
  private _offset: Vector = new Vector(0, 0);
  private _scale: number = 1; // 1 = 100%, [1..3]

  get offset() {
    return this._offset;
  }

  get scale() {
    return this._scale;
  }

  set offset(offset: Vector) {
    this._offset = offset;
  }

  set scale(scale: number) {
    this._scale = clamp(scale, 0.75, 2);
  }

  constructor() {
    makeAutoObservable(this);
  }

  resetOffsetAndScale() {
    this.offset = new Vector(0, 0);
    this.scale = 1;
  }
}