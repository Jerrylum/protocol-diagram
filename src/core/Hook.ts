import React from "react";

export function useBetterMemo<T extends { destructor: () => void } | {}>(
  factory: () => T,
  deps: React.DependencyList = []
): T {
  const storage = React.useMemo(() => factory(), deps); // eslint-disable-line react-hooks/exhaustive-deps
  React.useEffect(() => {
    return () => {
      if ("destructor" in storage && typeof storage.destructor === "function") storage.destructor?.();
    };
  }, deps); // eslint-disable-line react-hooks/exhaustive-deps

  return storage;
}
