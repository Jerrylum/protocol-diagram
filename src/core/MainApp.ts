import { CancellableCommand } from "../command/Commands";
import { Diagram, Memento } from "../diagram/Diagram";
import { Timeline } from "../diagram/Timeline";

export class MainApp extends Timeline<CancellableCommand> {
  // The current memento saved in the file or the first memento in the timeline
  private sourceCurrentMemento!: Memento | null;
  private isModifiedFlag!: boolean;

  constructor() {
    super(new Diagram());
    this.newDiagram();
  }

  get diagram(): Diagram {
    return this._diagram;
  }

  set diagram(diagram: Diagram) {
    this._diagram = diagram;
  }

  newDiagram() {
    this.diagram = new Diagram();
    this.resetHistory();
  }
  
  /**
   * a method that checks whether the diagram is modified, returns true if the
   * flag `isModified` is true or the file is not saved
   *
   * @return whether the diagram is modified
   */
  isModified(): boolean {
    return this.isModifiedFlag || this.sourceCurrentMemento !== this.getLatestMemento();
  }

  /**
   * a method that sets the flag `isModified` to the given value., note that
   * setting the flag to false does not mean the diagram is not modified
   *
   * @see #isModified()
   *
   * @param isModified the value to set
   */
  setModified(isModified: boolean) {
    this.isModifiedFlag = isModified;
  }

  /**
   * a method that eliminates all history and set the flag `isModified` to false
   */
  resetHistory() {
    super.resetHistory();
    this.sourceCurrentMemento = this.getLatestMemento();
    this.isModifiedFlag = false;
  }

  // diagram: Diagram = new Diagram();
  // handler: MainDiagramHandler = new MainDiagramHandler(this);
}
