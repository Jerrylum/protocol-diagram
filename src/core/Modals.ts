import { makeAutoObservable } from "mobx";

export class Modals {
  private opening_: {
    symbol: Symbol;
    priority: number;
  } | null = null;

  get opening(): Symbol | null {
    return this.opening_?.symbol ?? null;
  }

  get priority(): number | null {
    return this.opening_?.priority ?? null;
  }

  get isOpen() {
    return this.opening_ !== null;
  }

  open(symbol: Symbol, priority: number = 0): boolean {
    if (this.opening_ === null || this.opening_.priority <= priority) {
      this.opening_ = { symbol: symbol, priority };
      return true;
    } else {
      return false;
    }
  }

  close(symbol?: Symbol) {
    if (symbol === undefined || this.opening_?.symbol === symbol) this.opening_ = null;
  }

  constructor() {
    makeAutoObservable(this);
  }
}

