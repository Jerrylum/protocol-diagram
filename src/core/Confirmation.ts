import { makeAutoObservable, when } from "mobx";
import { Modals } from "./Modals";

export interface ConfirmationButton {
  label: string;
  onClick?: () => void;
  hotkey?: string;
  color?: "inherit" | "primary" | "secondary" | "success" | "error" | "info" | "warning";
}

export interface ConfirmationPromptData {
  title: string;
  description: React.ReactNode;
  buttons: ConfirmationButton[];
  onKeyDown?(e: React.KeyboardEvent<HTMLDivElement>, onClick: (index: number) => void): void;
}

export interface ConfirmationInputPromptData extends ConfirmationPromptData {
  inputLabel: string;
  inputDefaultValue: string;
}

export class Confirmation {
  private data?: ConfirmationPromptData | ConfirmationInputPromptData;
  public input?: string;

  constructor(private modals: () => Modals) {
    makeAutoObservable(this);
  }

  close() {
    this.modals().close(ConfirmationModalSymbol);
    this.data = undefined;
  }

  async prompt(data: ConfirmationPromptData | ConfirmationInputPromptData, priority: number = 1) {
    if (data.buttons.length === 0) {
      data.buttons.push({ label: "OK" });
    }

    this.close();
    this.data = data;
    if ("inputLabel" in data && "inputDefaultValue" in data) {
      this.input = data.inputDefaultValue;
    } else {
      this.input = undefined;
    }

    this.modals().open(ConfirmationModalSymbol, priority);

    await when(() => this.data === undefined);

    return this.input;
  }

  get isOpen() {
    return this.data !== undefined && this.modals().opening === ConfirmationModalSymbol;
  }

  get title() {
    return this.data?.title ?? "";
  }

  set title(value: string) {
    if (this.data === undefined) return;
    this.data.title = value;
  }

  get description() {
    return this.data?.description ?? "";
  }

  set description(value: React.ReactNode) {
    if (this.data === undefined) return;
    this.data.description = value;
  }

  get buttons() {
    return this.data?.buttons ?? [];
  }

  get onKeyDown() {
    return this.data?.onKeyDown;
  }

  get inputLabel() {
    return (this.data as ConfirmationInputPromptData)?.inputLabel ?? "";
  }
}

export const ConfirmationModalSymbol = Symbol("ConfirmationModalSymbol");
