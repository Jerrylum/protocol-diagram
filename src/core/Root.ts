import { Modals } from "./Modals";

export type RootStore = typeof rootStore;

const rootStore = {
  modals: new Modals()
} as const;

export function getRootStore(): RootStore {
  return rootStore;
}

