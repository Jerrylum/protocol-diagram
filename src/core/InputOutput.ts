import { enqueueErrorSnackbar, enqueueSuccessSnackbar } from "../app/Notice";
import { Logger } from "./Logger";
import { IOFileHandle } from "./MainApp";
import { getRootStore } from "./Root";
import { isFirefox, runInActionAsync } from "./Util";

const logger = Logger("I/O");

async function saveConfirm(callback: () => void) {
  const { app, confirmation } = getRootStore();

  return new Promise<boolean>((resolve, reject) => {
    confirmation.prompt({
      title: "Unsaved Changes",
      description: "Do you want to save the changes made to " + app.mountingFile.name + "?",
      buttons: [
        {
          label: "Save",
          color: "success",
          hotkey: "s",
          onClick: async () => {
            if (await onSave()) {
              callback();
              resolve(true);
            } else {
              resolve(false);
            }
          }
        },
        {
          label: "Don't Save",
          hotkey: "n",
          onClick: () => {
            callback();
            resolve(true);
          }
        },
        { label: "Cancel", onClick: () => resolve(false) }
      ]
    });
  });
}

async function fileNameConfirm(description: string, callback: () => void) {
  const { app, confirmation } = getRootStore();

  return new Promise<void>((resolve, reject) => {
    confirmation.prompt({
      title: "Download",
      description,
      buttons: [
        {
          label: "Confirm",
          color: "success",
          onClick: async () => {
            let candidate = confirmation.input ?? app.mountingFile.name;
            if (candidate.indexOf(".") === -1) candidate += ".json";
            app.mountingFile.name = candidate;
            app.mountingFile.isNameSet = true;
            callback();
            resolve();
          }
        },
        { label: "Cancel", onClick: () => resolve() }
      ],
      inputLabel: "File Name",
      inputDefaultValue: app.mountingFile.name
    });
  });
}

async function writeFile(contents: string): Promise<boolean> {
  const { app } = getRootStore();

  try {
    const file = app.mountingFile;
    if (file.handle === null) throw new Error("fileHandle is undefined");

    // XXX
    await file.handle.requestPermission({ mode: "readwrite" });

    const writable = await file.handle.createWritable();
    await writable.write(contents);
    await writable.close();

    enqueueSuccessSnackbar(logger, "Saved");
    return true;
  } catch (err) {
    if (err instanceof DOMException) enqueueErrorSnackbar(logger, "Failed to save file");
    else enqueueErrorSnackbar(logger, err);
    return false;
  }
}

async function readFile(): Promise<string | undefined> {
  const { app } = getRootStore();

  const options = {
    types: [{ description: "Protocol Diagram", accept: { "application/json": [] } }],
    excludeAcceptAllOption: false,
    multiple: false
  };

  try {
    const [fileHandle] = await showOpenFilePicker(options);
    app.mountingFile.handle = fileHandle;
    app.mountingFile.name = fileHandle.name;
    app.mountingFile.isNameSet = true;

    const file = await fileHandle.getFile();
    const contents = await file.text();

    return contents;
  } catch (err) {
    if (err instanceof DOMException === false) enqueueErrorSnackbar(logger, err);
    else logger.error(err); // UX: Do not show DOMException to user, usually means user cancelled

    return undefined;
  }
}

async function readFileFromInput(): Promise<string | undefined> {
  const { app } = getRootStore();

  const input = document.createElement("input");
  input.type = "file";
  input.multiple = false;
  input.accept = "application/json";

  // See https://stackoverflow.com/questions/47664777/javascript-file-input-onchange-not-working-ios-safari-only
  Object.assign(input.style, { position: "fixed", top: "-100000px", left: "-100000px" });

  document.body.appendChild(input);

  await new Promise(resolve => {
    input.addEventListener("change", resolve, { once: true });
    input.click();
  });

  // ALGO: Remove polyfill input[type=file] elements, including elements from last time
  document.querySelectorAll("body > input[type=file]").forEach(input => input.remove());

  const file = input.files?.[0];
  if (file === undefined) return undefined;

  await runInActionAsync(() => {
    app.mountingFile.handle = null;
    app.mountingFile.name = file.name;
    app.mountingFile.isNameSet = true;
  });

  const reader = new FileReader();
  reader.readAsText(file);

  return new Promise<string | undefined>((resolve, reject) => {
    reader.onload = () => resolve(reader.result as string);
    reader.onerror = () => resolve(undefined);
  });
}

function downloadFile(contents: string) {
  const { app } = getRootStore();

  const a = document.createElement("a");
  const file = new Blob([contents], { type: "application/json" });
  a.href = URL.createObjectURL(file);
  a.download = app.mountingFile.name;
  a.click();
}

async function choiceSave(): Promise<boolean> {
  const { app } = getRootStore();

  const options = {
    types: [{ description: "Protocol Diagram", accept: { "application/json": [] } }],
    suggestedName: app.mountingFile.name,
    excludeAcceptAllOption: false,
    multiple: false
  };

  try {
    const fileHandle = await window.showSaveFilePicker(options);
    app.mountingFile.handle = fileHandle;
    app.mountingFile.name = fileHandle.name;
    app.mountingFile.isNameSet = true;

    return true;
  } catch (err) {
    logger.error(err); // ignore error
    return false;
  }
}

export function isFileSystemSupported() {
  return window.showOpenFilePicker !== undefined && window.showSaveFilePicker !== undefined;
}

export async function onNew(saveCheck: boolean = true): Promise<boolean> {
  const { app } = getRootStore();

  if (saveCheck && app.isModified()) return saveConfirm(onNew.bind(null, false));

  app.newDiagram();
  app.mountingFile = new IOFileHandle();
  return true;
}

export async function onSave(): Promise<boolean> {
  const { app } = getRootStore();

  if (isFileSystemSupported() === false) return onDownload(true);

  if (app.mountingFile.handle === null) return onSaveAs();

  const output = app.exportDiagram();

  if (await writeFile(output)) {
    app.save();
    return true;
  } else {
    return false;
  }
}

export async function onSaveAs(): Promise<boolean> {
  const { app } = getRootStore();

  if (isFileSystemSupported() === false) return onDownloadAs(true);

  const output = app.exportDiagram();

  if (!(await choiceSave())) return false;

  if (await writeFile(output)) {
    app.save();
    return true;
  } else {
    return false;
  }
}

export async function onOpen(saveCheck: boolean = true, interactive: boolean = true): Promise<boolean> {
  const { app, confirmation } = getRootStore();

  if (saveCheck && app.isModified()) return saveConfirm(onOpen.bind(null, false, false));

  if (interactive && isFirefox()) {
    // Resolve: <input> picker was blocked due to lack of user activation.
    await confirmation.prompt({
      title: "Open File",
      description: "Press any key to continue.",
      buttons: [{ label: "Open", color: "success" }],
      onKeyDown: (e, onClick) => onClick(0)
    });
  }

  const contents = await (isFileSystemSupported() ? readFile() : readFileFromInput());
  if (contents === undefined) return false;

  try {
    await app.importDiagram(contents);
    return true;
  } catch (err) {
    enqueueErrorSnackbar(logger, err);
    return false;
  }
}

export async function onDownload(fallback: boolean = false): Promise<boolean> {
  const { app } = getRootStore();

  if (app.mountingFile.isNameSet === false) return onDownloadAs(fallback);

  const output = app.exportDiagram();

  downloadFile(output);

  return true;
}

export async function onDownloadAs(fallback: boolean = false): Promise<boolean> {
  const { app } = getRootStore();

  const output = app.exportDiagram();

  fileNameConfirm(
    fallback ? "Writing file to the disk is not supported in this browser. Falling back to download." : "",
    downloadFile.bind(null, output)
  );

  return true;
}

export async function onDropFile(file: File, saveCheck: boolean = true): Promise<boolean> {
  const { app } = getRootStore();

  if (saveCheck && app.isModified()) return saveConfirm(onDropFile.bind(null, file, false));

  await runInActionAsync(() => {
    app.mountingFile.handle = null;
    app.mountingFile.name = file.name;
    app.mountingFile.isNameSet = true;
  });

  const reader = new FileReader();
  reader.readAsText(file);

  const contents = await new Promise<string | undefined>((resolve, reject) => {
    reader.onload = () => resolve(reader.result as string);
    reader.onerror = () => resolve(undefined);
  });
  if (contents === undefined) return false;

  try {
    await app.importDiagram(contents);
    return true;
  } catch (err) {
    enqueueErrorSnackbar(logger, err);
    return false;
  }
}

