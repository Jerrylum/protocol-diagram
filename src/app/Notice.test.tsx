import { NoticeProvider, enqueueSuccessSnackbar, enqueueInfoSnackbar, enqueueErrorSnackbar } from "./Notice";
import { render } from "@testing-library/react";
import { Logger } from "../core/Logger";

test("Render NoticeProvider", () => {
  const components = (
    <div id="root-container">
      <NoticeProvider />
    </div>
  );
  const result = render(components);
  enqueueErrorSnackbar(Logger("testError"), "testError", 5000);
  enqueueSuccessSnackbar(Logger("testSuccess"), "testSuccess", 5000);
  enqueueInfoSnackbar(Logger("testInfo"), "testInfo", 5000);
  result.rerender(components);
  expect(result.container).toMatchSnapshot();

  const barError = result.container.querySelectorAll("#notistack-snackbar")[0];
  const barSuccess = result.container.querySelectorAll("#notistack-snackbar")[1];
  const barInfo = result.container.querySelectorAll("#notistack-snackbar")[2];

  expect(barError.textContent).toBe("testError");
  expect(barSuccess.textContent).toBe("testSuccess");
  expect(barInfo.textContent).toBe("testInfo");
});
