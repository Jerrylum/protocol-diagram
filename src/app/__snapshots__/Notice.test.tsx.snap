// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render NoticeProvider 1`] = `
<div>
  <div
    id="root-container"
  >
    <div
      class="notistack-SnackbarContainer go1453831412 go4027089540 go1141946668"
    >
      <div
        class="go1475592160 go1671063245"
        style="pointer-events: all; overflow: visible; min-height: 0px; transition: height 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      >
        <div
          class="notistack-CollapseWrapper"
          style="display: flex; width: 100%;"
        >
          <div
            class="notistack-Snackbar go3963613292"
          >
            <div
              aria-describedby="notistack-snackbar"
              class="go1888806478 notistack-MuiContent notistack-MuiContent-error go167266335 go3651055292  css-6ods1o go3162094071"
              role="alert"
              style="webkit-transform: none; transform: none; webkit-transition: -webkit-transform 225ms cubic-bezier(0.0, 0, 0.2, 1) 0ms; transition: transform 225ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;"
            >
              <div
                class="go946087465"
                id="notistack-snackbar"
              >
                <svg
                  focusable="false"
                  style="font-size: 20px; margin-inline-end: 8px; user-select: none; width: 1em; height: 1em; display: inline-block; fill: currentColor; flex-shrink: 0;"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12,2C17.53,2 22,6.47 22,12C22,17.53 17.53,22 12,22C6.47,22 2,17.53 2,12C2,
        6.47 6.47,2 12,2M15.59,7L12,10.59L8.41,7L7,8.41L10.59,12L7,15.59L8.41,17L12,
        13.41L15.59,17L17,15.59L13.41,12L17,8.41L15.59,7Z"
                  />
                </svg>
                testError
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="go1475592160 go1671063245"
        style="pointer-events: all; overflow: visible; min-height: 0px; transition: height 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      >
        <div
          class="notistack-CollapseWrapper"
          style="display: flex; width: 100%;"
        >
          <div
            class="notistack-Snackbar go3963613292"
          >
            <div
              aria-describedby="notistack-snackbar"
              class="go1888806478 notistack-MuiContent notistack-MuiContent-success go167266335 go1725278324  css-6ods1o go3162094071"
              role="alert"
              style="webkit-transform: none; transform: none; webkit-transition: -webkit-transform 225ms cubic-bezier(0.0, 0, 0.2, 1) 0ms; transition: transform 225ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;"
            >
              <div
                class="go946087465"
                id="notistack-snackbar"
              >
                <svg
                  focusable="false"
                  style="font-size: 20px; margin-inline-end: 8px; user-select: none; width: 1em; height: 1em; display: inline-block; fill: currentColor; flex-shrink: 0;"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 2C6.5 2 2 6.5 2 12S6.5 22 12 22 22 17.5 22 12 17.5 2 12 2M10 17L5 12L6.41
        10.59L10 14.17L17.59 6.58L19 8L10 17Z"
                  />
                </svg>
                testSuccess
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="go1475592160 go1671063245"
        style="pointer-events: all; overflow: visible; min-height: 0px; transition: height 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      >
        <div
          class="notistack-CollapseWrapper"
          style="display: flex; width: 100%;"
        >
          <div
            class="notistack-Snackbar go3963613292"
          >
            <div
              aria-describedby="notistack-snackbar"
              class="go1888806478 notistack-MuiContent notistack-MuiContent-info go167266335 go1930647212 go3162094071"
              role="alert"
              style="webkit-transform: none; transform: none; webkit-transition: -webkit-transform 225ms cubic-bezier(0.0, 0, 0.2, 1) 0ms; transition: transform 225ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;"
            >
              <div
                class="go946087465"
                id="notistack-snackbar"
              >
                <svg
                  focusable="false"
                  style="font-size: 20px; margin-inline-end: 8px; user-select: none; width: 1em; height: 1em; display: inline-block; fill: currentColor; flex-shrink: 0;"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M13,9H11V7H13M13,17H11V11H13M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,
        0 22,12A10,10 0 0,0 12,2Z"
                  />
                </svg>
                testInfo
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
